[Clang][analyzer] replace Stmt* with ConstCFGElementRef in SymbolConjured

fangyi-zhou opened the pull request:
Closes #57270.

This PR changes the `Stmt *` field in `SymbolConjured` with `CFGBlock::ConstCFGElementRef`. The motivation is that, when conjuring a symbol, there might not always be a statement available, causing information to be lost for conjured symbols, whereas the CFGElementRef can always be provided at the callsite.

Following the idea, this PR changes callsites of functions to create conjured symbols, and replaces them with appropriate `CFGElementRef`s.

fangyi-zhou commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SValBuilder.h:
```diff
@@ -171,20 +172,27 @@ class SValBuilder {
-                                      const void *symbolTag = nullptr) {
-    return SymMgr.conjureSymbol(expr, LCtx, visitCount, symbolTag);
-  }
+  // const SymbolConjured* conjureSymbol(const Stmt *stmt,
```
Will remove these commented out lines before marking this PR as ready.

fangyi-zhou commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -113,19 +120,21 @@ class SymbolConjured : public SymbolData {
                       const void *SymbolTag) {
     profile.AddInteger((unsigned)SymbolConjuredKind);
-    profile.AddPointer(S);
+    // profile.Add(ElemRef);
```
Q: Do I need to add `ElemRef` to the profile?

isuckatcs replied:
Yes

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/RegionStore.cpp:
```diff
@@ -1376,8 +1379,8 @@ StoreRef RegionStoreManager::invalidateRegions(
   RegionBindingsRef B = getRegionBindings(store);
-  InvalidateRegionsWorker W(*this, StateMgr, B, S, Count, LCtx, IS, ITraits,
-                            Invalidated, GlobalsFilter);
+  InvalidateRegionsWorker W(*this, StateMgr, B, Call->getCFGElementRef(), Count,
```
Q: Here `Call` is possibly null, how should I get a ref here?

isuckatcs replied:
Here we initially passed `const Stmt *S` to the worker, so we should pass the `CFGElementRef` of `const Stmt *S` and not what we obtain from the call, I guess.

As for that, you can pass it down from the only callsite of this function, as you have access to the engine.
```c++
ProgramStateRef ProgramState::invalidateRegions(
  ...
  ExprEngine &Eng = Mgr.getOwningEngine()
  ...
  const StoreRef &NewStore = Mgr.StoreMgr->invalidateRegions(
      getStore(), Values, Eng.getCFGElementRef(), Count, LCtx, Call, *IS,
      *ITraits, &TopLevelInvalidated, &Invalidated);
  ...
}
```

I don't know if that's a reference that we want here, but with trial and error we will see.

fangyi-zhou replied:
Are you referring to this instance that I should pass the refernece to the statement instead of the call? I think I misunderstood this comment so I changed the references to not use the one from the call in all places.

isuckatcs commented:
The source of the crash you mentioned in the issue is `CStringChecker.cpp:1304`, where the `CallEvent` is a `nullptr`.
```c++
return State->invalidateRegions(R, E, C.blockCount(), LCtx,
                                CausesPointerEscape, nullptr, nullptr,
                                &ITraits);
```

isuckatcs commented on clang/lib/StaticAnalyzer/Core/SymbolManager.cpp:
```diff
@@ -82,7 +82,7 @@ void UnarySymExpr::dumpToStream(raw_ostream &os) const {
 void SymbolConjured::dumpToStream(raw_ostream &os) const {
   os << getKindStr() << getSymbolID() << '{' << T << ", LC" << LCtx->getID();
-  if (S)
+  if (const auto *S = getStmt())
```
Here we might actually want to dump the `CFGElementRef` not the statement.

isuckatcs reviewed (changes requested):
So far we have some progress, so keep it up.

Keep in mind that we'll also need to update `SValExplainer`, but you'll see it once you run the tests and start seeing the warning messages.

fangyi-zhou commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -111,8 +111,13 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
+    llvm::raw_string_ostream OS(Str);
+    OS << "symbol of type '" + S->getType().getAsString() +
+              "' conjured at statement '";
+    S->getCFGElementRef()->dumpToStream(OS);
```
Q: Do we have a way to pretty print a CFGElementRef instead of `dumpToStream`?

isuckatcs replied:
I don't know about any. You could create a similar helper to `printStmt()` in this class though.

isuckatcs replied:
Yes, but you can extract the "dumping to an ostream" logic into a separate function and tweak it there however you want. This is what I ment by similar to `printStmt()`.

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/SValBuilder.cpp:
```diff
@@ -151,72 +151,63 @@ SValBuilder::getRegionValueSymbolVal(const TypedValueRegion *region) {
-
-  return conjureSymbolVal(SymbolTag, Ex, LCtx, T, Count);
-}
+// DefinedOrUnknownSVal SValBuilder::conjureSymbolVal(const void *SymbolTag,
```
Note to self: This should probably be added back in some shape or form since it has some different behaviour when handling expressions with a provided type --- which means it's not entirely same to calling `conjureSymbolVal(SymbolTag, Ex, LCtx, Ex->getType(), Count)`. Probably need to revisit all callsites and check.

fangyi-zhou commented:
I've made some more progress, the crash goes away, there are still some review comments that I need to address, which I'll try to complete later.

```
/home/fangyi/playground/bug.cc:21:5: warning: value derived from (symbol of type 'int' conjured at statement '->~S() (Implicit destructor)
') for global variable 'S::a' [debug.ExprInspection]
   21 |     clang_analyzer_explain(S::a);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

isuckatcs commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -113,8 +113,18 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
+    S->getCFGElementRef()->dumpToStream(OS);
+    // HACK: dumpToStream will output a new line in the end, this is
+    // undesireable, thus we remove it.
+    if (Str.back() == '\n') {
```
What if we are on windows and the line ending is `\r\n`?

isuckatcs commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -113,8 +113,18 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
+    if (Str.back() == '\n') {
+      Str.pop_back();
+    }
+    OS << "'";
```
```suggestion
    OS << '\'';
```

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h:
```diff
@@ -151,6 +151,11 @@ class CheckerContext {
     return Pred->getSVal(S);
   }
 
+  /// Get the CFG Element Ref from the ExprEngine
```
I think you can remove this comment. It's obvious what the function does even without it.

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SValBuilder.h:
```diff
@@ -171,19 +172,11 @@ class SValBuilder {
-                                      const void *symbolTag = nullptr) {
-    return SymMgr.conjureSymbol(expr, LCtx, visitCount, symbolTag);
+  const SymbolConjured *
+  conjureSymbol(const CFGBlock::ConstCFGElementRef ElemRef,
```
Here we had an overload without a `QualType`, don't we need it?

Xazax-hun replied:
Could we still extract a type from the `ElemRef` in the implementation? Does the caller have a reasonable type to pass in when `ElemRef` is not referring to an expression?

fangyi-zhou replied:
We can try to downcast the `ElemRef` for a statement, and try to downcast as an expression to obtain the type in those cases. I guess it would work in the callsites in the deleted overload if the `ElemRef` are statements.

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SValBuilder.h:
```diff
@@ -198,32 +191,24 @@ class SValBuilder {
-                                        QualType type,
-                                        unsigned visitCount);
+  DefinedOrUnknownSVal
+  conjureSymbolVal(const void *symbolTag, const Expr *expr,
+                   const CFGBlock::ConstCFGElementRef elemRef,
```
Why do we need both a `const Expr *` and a `const CFGBlock::ConstCFGElementRef`? I'd imagine, the `ConstCFGElementRef` points to the `Expr`.

fangyi-zhou replied:
If you look at the implementation code, there's some special treatment wrt the type of the expression

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SValBuilder.h:
```diff
@@ -198,32 +191,24 @@ class SValBuilder {
-                                       const LocationContext *LCtx,
-                                       QualType type, unsigned Count);
+  DefinedSVal
+  getConjuredHeapSymbolVal(const CFGBlock::ConstCFGElementRef elemRef,
```
The overload without a `QualType` is missing. Why did we remove it?

fangyi-zhou replied:
Same as above.

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -80,17 +81,18 @@ class SymbolRegionValue : public SymbolData {
+                 const void *symbolTag)
+      : SymbolData(SymbolConjuredKind, sym), ElemRef(elemRef), T(t),
+        Count(count), LCtx(lctx), SymbolTag(symbolTag) {
     // FIXME: 's' might be a nullptr if we're conducting invalidation
```
This FIXME is no longer valid, because a `ConstCFGElementRef` is never a `nullptr`, IIUC.

isuckatcs commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -533,18 +538,12 @@ class SymbolManager {
-    return acquire<SymbolConjured>(E, LCtx, T, VisitCount, SymbolTag);
-  }
+  const SymbolConjured *
+  conjureSymbol(const CFGBlock::ConstCFGElementRef ElemRef,
```
An overload without a type is missing.

fangyi-zhou replied:
Same as above.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -1299,8 +1299,8 @@ ProgramStateRef CStringChecker::invalidateBufferAux(
     RegionAndSymbolInvalidationTraits ITraits;
     bool CausesPointerEscape = InvalidationTraitOperations(ITraits, R);
 
-    return State->invalidateRegions(R, E, C.blockCount(), LCtx,
```
This change makes `E` an unused parameter.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -1515,7 +1515,8 @@ void CStringChecker::evalCopyCommon(CheckerContext &C, const CallEvent &Call,
       if (lastElement.isUnknown())
         lastElement = C.getSValBuilder().conjureSymbolVal(
-            nullptr, Call.getOriginExpr(), LCtx, C.blockCount());
+            nullptr, Call.getOriginExpr(), C.getCFGElementRef(), LCtx,
```
At this point, isn't it the `Call` that conjured the symbol? Shouldn't we pass a `CFGElementRef` to the `Call` instead?

fangyi-zhou replied:
I don't know what's the correct one to pass and don't have a way to check reliably since I don't understand the static analysis code. I don't understand your intention in this comment, do you mean I should pass the CFGElementRef to the Call or not?

Xazax-hun replied:
You can dump the CFGElementRef and the expr. We want to make sure they match, so there is no semantic change. That is one way to check if the code is correct.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -1665,8 +1666,9 @@ void CStringChecker::evalMemcmp(CheckerContext &C, const CallEvent &Call,
-      SVal CmpV = Builder.conjureSymbolVal(nullptr, Call.getOriginExpr(), LCtx,
-                                           C.blockCount());
+      SVal CmpV =
+          Builder.conjureSymbolVal(nullptr, Call.getOriginExpr(),
+                                   C.getCFGElementRef(), LCtx, C.blockCount());
```
Same here

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -1770,7 +1772,8 @@ void CStringChecker::evalstrLengthCommon(CheckerContext &C,
       // and the limit. This is better than nothing.
       result = C.getSValBuilder().conjureSymbolVal(
-          nullptr, Call.getOriginExpr(), LCtx, C.blockCount());
+          nullptr, Call.getOriginExpr(), C.getCFGElementRef(), LCtx,
+          C.blockCount());
```
And here.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -1794,7 +1797,8 @@ void CStringChecker::evalstrLengthCommon(CheckerContext &C,
     if (result.isUnknown()) {
       result = C.getSValBuilder().conjureSymbolVal(
-          nullptr, Call.getOriginExpr(), LCtx, C.blockCount());
+          nullptr, Call.getOriginExpr(), C.getCFGElementRef(), LCtx,
+          C.blockCount());
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2261,7 +2265,8 @@ void CStringChecker::evalStrcpyCommon(CheckerContext &C, const CallEvent &Call,
     // overflow, we still need a result. Conjure a return value.
     if (ReturnEnd && Result.isUnknown()) {
-      Result = svalBuilder.conjureSymbolVal(nullptr, Call.getOriginExpr(), LCtx,
+      Result = svalBuilder.conjureSymbolVal(nullptr, Call.getOriginExpr(),
+                                            C.getCFGElementRef(), LCtx,
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2361,8 +2366,9 @@ void CStringChecker::evalStrcmpCommon(CheckerContext &C, const CallEvent &Call,
-  SVal resultVal = svalBuilder.conjureSymbolVal(nullptr, Call.getOriginExpr(),
-                                                LCtx, C.blockCount());
+  SVal resultVal =
+      svalBuilder.conjureSymbolVal(nullptr, Call.getOriginExpr(),
+                                   C.getCFGElementRef(), LCtx, C.blockCount());
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2469,14 +2475,14 @@ void CStringChecker::evalStrsep(CheckerContext &C,
     // Conjure a symbolic value. It's the best we can do.
-    Result = SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(), LCtx,
-                                  C.blockCount());
+    Result = SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(),
+                                  C.getCFGElementRef(), LCtx, C.blockCount());
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2469,14 +2475,14 @@ void CStringChecker::evalStrsep(CheckerContext &C,
     State =
         State->bindLoc(*SearchStrLoc,
-                       SVB.conjureSymbolVal(getTag(), Call.getOriginExpr(),
+                       SVB.conjureSymbolVal(getTag(), C.getCFGElementRef(),
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2520,7 +2526,8 @@ void CStringChecker::evalStdCopyCommon(CheckerContext &C,
 
   SVal ResultVal =
-      SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(), LCtx, C.blockCount());
+      SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(), C.getCFGElementRef(),
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/ContainerModeling.cpp:
```diff
@@ -326,7 +330,7 @@ void ContainerModeling::handleAssignment(CheckerContext &C, SVal Cont,
           auto &SVB = C.getSValBuilder();
           // Then generate and assign a new "end" symbol for the new container.
           auto NewEndSym =
-              SymMgr.conjureSymbol(CE, C.getLocationContext(),
```
This makes `CE` an unused parameter.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/ErrnoModeling.cpp:
```diff
@@ -260,7 +260,7 @@ ProgramStateRef setErrnoStdMustBeChecked(ProgramStateRef State,
   const MemRegion *ErrnoR = State->get<ErrnoRegion>();
   if (!ErrnoR)
     return State;
-  State = State->invalidateRegions(ErrnoR, InvalE, C.blockCount(),
```
`InvalE` is an unused parameter, the element reference should point to that I think.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/ErrnoTesterChecker.cpp:
```diff
@@ -130,8 +130,9 @@ void ErrnoTesterChecker::evalSetErrnoIfErrorRange(CheckerContext &C,
-  DefinedOrUnknownSVal ErrnoVal = SVB.conjureSymbolVal(
-      nullptr, Call.getOriginExpr(), C.getLocationContext(), C.blockCount());
+  DefinedOrUnknownSVal ErrnoVal =
+      SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(), C.getCFGElementRef(),
```
Same as multiple previous question about the call.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/IteratorModeling.cpp:
```diff
@@ -467,7 +467,7 @@ void IteratorModeling::handleComparison(CheckerContext &C, const Expr *CE,
   SymbolRef Sym;
   if (!LPos || !RPos) {
     auto &SymMgr = C.getSymbolManager();
-    Sym = SymMgr.conjureSymbol(CE, C.getLocationContext(),
```
`CE` is unused parameter.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/IteratorModeling.cpp:
```diff
@@ -494,7 +494,7 @@ void IteratorModeling::handleComparison(CheckerContext &C, const Expr *CE,
     auto &SymMgr = C.getSymbolManager();
     auto *LCtx = C.getLocationContext();
     RetVal = nonloc::SymbolVal(SymMgr.conjureSymbol(
-        CE, LCtx, C.getASTContext().BoolTy, C.blockCount()));
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/IteratorModeling.cpp:
```diff
@@ -688,7 +688,8 @@ void IteratorModeling::assignToContainer(CheckerContext &C, const Expr *CE,
 
   auto State = C.getState();
   const auto *LCtx = C.getLocationContext();
-  State = createIteratorPosition(State, RetVal, Cont, CE, LCtx, C.blockCount());
```
.

isuckatcs commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -229,7 +230,7 @@ DefinedSVal makeRetVal(CheckerContext &C, const CallExpr *CE) {
 
   const LocationContext *LCtx = C.getLocationContext();
   return C.getSValBuilder()
-      .conjureSymbolVal(nullptr, CE, LCtx, C.blockCount())
```
`CE` in unused.

isuckatcs reviewed (changes requested):
I didn't have enough time to check the whole patch, I'll get back to it later.

The general patterns seems to be that when a conjured symbol is created, it's always the reference to the current CFG element that is passed as the source statement instead of some other statement that was passed in the past. 

I think every place, where we conjure a symbol should be double checked, so that we make sure we pass the CFG element reference to the correct statement, that actually conjured the value.

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/SValBuilder.cpp:
```diff
@@ -166,57 +166,47 @@ DefinedOrUnknownSVal SValBuilder::conjureSymbolVal(const void *SymbolTag,
   return nonloc::SymbolVal(sym);
 }
 
-DefinedSVal SValBuilder::getConjuredHeapSymbolVal(const Expr *E,
```
This overload is removed since it's not possible to retrieve the type from the CFGElementRef. If there's a way to obtain it please let me know so I can add it back.

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/SValBuilder.cpp:
```diff
@@ -151,10 +151,10 @@ SValBuilder::getRegionValueSymbolVal(const TypedValueRegion *region) {
   return nonloc::SymbolVal(sym);
 }
 
-DefinedOrUnknownSVal SValBuilder::conjureSymbolVal(const void *SymbolTag,
```
This overload would have been removed, but its behaviour is not entirely the same as obtaining the type from the expression directly. (See the check for `isGLValue`). Without complicating the refactor in this PR I decided to leave it as is, and add a new parameter for the CFG Element Ref

Xazax-hun commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/LoopWidening.h:
```diff
@@ -27,7 +27,8 @@ namespace ento {
 ProgramStateRef getWidenedLoopState(ProgramStateRef PrevState,
                                     const LocationContext *LCtx,
-                                    unsigned BlockCount, const Stmt *LoopStmt);
+                                    unsigned BlockCount, const Stmt *LoopStmt,
```
Do we need both the Stmt and the ElemRef here?

fangyi-zhou replied:
Not any more, will remove.

Xazax-hun commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -101,8 +103,10 @@ class SymbolConjured : public SymbolData {
 
 public:
-  /// It might return null.
-  const Stmt *getStmt() const { return S; }
```
Would it make sense to keep `getStmt`? It could still return null but some callers might find it useful.

fangyi-zhou commented on clang/lib/StaticAnalyzer/Checkers/IteratorModeling.cpp:
```diff
@@ -494,7 +494,7 @@ void IteratorModeling::handleComparison(CheckerContext &C, const Expr *CE,
     RetVal = nonloc::SymbolVal(SymMgr.conjureSymbol(
-        CE, LCtx, C.getASTContext().BoolTy, C.blockCount()));
+        C.getCFGElementRef(), LCtx, C.getASTContext().BoolTy, C.blockCount()));
     State = State->BindExpr(CE, LCtx, RetVal);
```
Used here.

isuckatcs commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -29,6 +31,18 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
+    ElemRef->dumpToStream(OS);
+    // HACK: `CFGBlock::ConstCFGElementRef::dumpToStream` contains a new line
+    // character in the end of the string, we don't want it so we remove it here.
+    while (!Str.empty() && std::isspace(Str.back())) {
```
`std::isspace` is dependant on the local, so I would advise against using it.

Also don't we have a builtin `trim()` function in LLVM, which can be used here?

isuckatcs commented on clang/lib/StaticAnalyzer/Core/SValBuilder.cpp:
```diff
@@ -151,10 +153,13 @@ SValBuilder::getRegionValueSymbolVal(const TypedValueRegion *region) {
-                                                   const Expr *Ex,
-                                                   const LocationContext *LCtx,
-                                                   unsigned Count) {
+/// When using this overload, the \p elemRef provided must be a \p CFGStmt.
```
This is asserted. The comment is not necessary I think.

steakhal replied:
I agree, let's drop this line.
```suggestion
```

isuckatcs reviewed (changes requested):
Can you please add a testcase with the snippetin the issue that used to crash?

fangyi-zhou commented on clang/include/clang/Analysis/CFG.h:
```diff
@@ -695,6 +695,21 @@ class CFGBlock {
+      ID.AddInteger(Index);
+    }
+
+    int64_t getID() const {
```
Any suggestion for replacement?

isuckatcs replied:
Why do we want to print the ID in the first place? It probably doesn't hold a value to our users anyway.

Either print a stament if there is any, or a source location, etc. The CFG is essentially the statements in the source file in execution order, so I imagine we always have a statement.

fangyi-zhou commented:
May I get a re-review for the changes please?

steakhal reviewed (changes requested):
I've only spotchecked. It looks correct at first glance.
I somewhat dislike that practically every file needs to include CFG.h; I'd advise revisiting this.

Do you intentionally take `CFGBlock::ConstCFGElementRef` as a const parameter? That doesn't seem to do anything. I'd rather not spell `const` unless there is a compelling reason to do so.

Then the final question, why is `ConstCFGElementRef` nested under the `CFGBlock`? It makes it spell every time, leading to noise. I'd revisit this, and possibly change its implementation to allow less mouthful spelling of this type.

fangyi-zhou committed:
[Clang][analyzer] replace Stmt* with ConstCFGElementRef in SymbolConjured

Closes #57270.

This PR changes the `Stmt *` field in `SymbolConjured` with
`CFGBlock::ConstCFGElementRef`. The motivation is that, when conjuring a
symbol, there might not always be a statement available, causing
information to be lost for conjured symbols, whereas the CFGElementRef
can always be provided at the callsite.

Following the idea, this PR changes callsites of functions to create
conjured symbols, and replaces them with appropriate `CFGElementRef`s.

fangyi-zhou commented:
Sorry I've been a bit busy with other things, just had some time to address the review comments. Please let me know if anything else needs changing

steakhal commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -29,6 +32,17 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
   ASTContext &ACtx;
   ProgramStateRef State;
 
+  std::string printCFGElementRef(CFGBlock::ConstCFGElementRef ElemRef) {
```
Having the `CFGBlock::ConstCFGElementRef` qualified name spelled every time could get messy really quick.
I didn't count how many times this is spelled throughout the PR, but it's gotta be a lot.

I've just checked that forward declaring is not really an option, and hoisting could be also quite a bit of work. However, we could at least have a type alias to avoid the `CFGBlock::` being spelled out every time.

Could you put `using ConstCFGElementRef = CFGBlock::ConstCFGElementRef;` after the definition of `CFGBlock`?

steakhal commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/ProgramState.h:
```diff
@@ -313,7 +313,8 @@ class ProgramState : public llvm::FoldingSetNode {
   ///
   /// \param Regions the set of regions to be invalidated.
-  /// \param E the expression that caused the invalidation.
+  /// \param ElemRef \p CFGBlock::ConstCFGElementRef that caused the
```
```suggestion
  /// \param ElemRef The CFG element that caused the
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -273,36 +273,41 @@ class CStringChecker : public Checker< eval::Call,
-                                    const Expr *BufE, SVal BufV, SVal SizeV,
-                                    QualType SizeTy);
+                                    const Expr *BufE,
+                                    CFGBlock::ConstCFGElementRef ElemRef,
```
In all of these examples, I feel that `ElemRef` is too verbose. `Elem` or `Place` should be a shorter name without compromising readability. Also, like I previously suggested, not having to spell out the `CFGBlock::` would be a huge help here too.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/CStringChecker.cpp:
```diff
@@ -2463,19 +2473,19 @@ void CStringChecker::evalStrsep(CheckerContext &C,
                                             LCtx, CharPtrTy, C.blockCount()),
                        LCtx);
   } else {
     assert(SearchStrVal.isUnknown());
     // Conjure a symbolic value. It's the best we can do.
-    Result = SVB.conjureSymbolVal(nullptr, Call.getOriginExpr(), LCtx,
+    Result = SVB.conjureSymbolVal(nullptr, Call.getCFGElementRef(), LCtx,
                                   C.blockCount());
   }
```
I think it's such a common pattern that we want to conjure at a call, that we should have a shorthand for it.
Conjure takes a bunch of obscured parameters and it's annoying to dig them up every time.
The "Call" should already have most stuff, thus we could have the opportunity to simplify the API for Calls.
It could be done later though. I just leave this for food for though.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/ErrnoModeling.h:
```diff
@@ -13,6 +13,7 @@
 #ifndef LLVM_CLANG_LIB_STATICANALYZER_CHECKERS_ERRNOMODELING_H
 #define LLVM_CLANG_LIB_STATICANALYZER_CHECKERS_ERRNOMODELING_H
 
+#include "clang/Analysis/CFG.h"
```
Actually, if we only declare an API that uses a class as a by-value parameter, that doesn't need the definition of the class.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/STLAlgorithmModeling.cpp:
```diff
@@ -97,11 +99,12 @@ bool STLAlgorithmModeling::evalCall(const CallEvent &Call,
   if (!Handler)
     return false;
 
-  return (this->**Handler)(C, CE);
+  return (this->**Handler)(C, CE, Call.getCFGElementRef());
```
Ideally, here we would pass the `Call` instead of `CE`. Then the handlers could dig up the Elem if they need that using the `Call.getCFGElementRef()`. I'd prefer this over what we propose here right now unless there is some compelling argument doing it this way.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -645,6 +645,7 @@ struct StreamOperationEvaluator {
   SymbolRef StreamSym = nullptr;
   const StreamState *SS = nullptr;
   const CallExpr *CE = nullptr;
+  std::optional<CFGBlock::ConstCFGElementRef> ElemRef;
```
Why is this declared here and not where the later assignment is?
As a rule of thumb, we should keep the lifetimes of objects as small as possible, and have their definitions as close to their uses as possible to enhance local reasoning.
Why do we need this optional anyway if we unconditionally unwrap it later?

fangyi-zhou replied:
I'm not entirely sure whether I get your point correctly. This value needs to be stored when `Init` is called, and used in other calls (e.g. `bindReturnValue`). The reason why this is declared as an optional is that there is no default constructor for `ConstCFGElementRef`, so I used an optional to give an absent value (I guess it could have been a pointer like other ones)

steakhal replied:
Ah I completely missed the context. nvm.

steakhal commented on clang/lib/StaticAnalyzer/Core/ExprEngineC.cpp:
```diff
@@ -21,18 +21,19 @@ using namespace ento;
 
 /// Optionally conjure and return a symbol for offset when processing
-/// an expression \p Expression.
+/// a \p CFGElementRef.
```
˙\p` should refer to a parameter name in the declaration. Consequently, this should be probably `ElemRef` currently.
But ideally, something like `Elem` or `Place` should be a better fit in general.

steakhal commented on clang/lib/StaticAnalyzer/Core/LoopWidening.cpp:
```diff
@@ -93,9 +76,8 @@ ProgramStateRef getWidenedLoopState(ProgramStateRef PrevState,
-  return PrevState->invalidateRegions(Regions, getLoopCondition(LoopStmt),
-                                      BlockCount, LCtx, true, nullptr, nullptr,
-                                      &ITraits);
+  return PrevState->invalidateRegions(Regions, ElemRef, BlockCount, LCtx, true,
```
I mean if it did so far, we didn't change the behavior, the assertion inside should still hold. Am I missing something?
That said, I wish we could keep the assertion within `getWidenedLoopState`.

fangyi-zhou replied:
I don't have a good answer to these questions... Any tips? I don't see any immediate ways to extract the required information

steakhal replied:
@isuckatcs

steakhal reviewed (changes requested):
Looks almost perfect. Thank you.
I've added a couple of comments, and suggestions.
There were a couple of suggestions that were not solved or commented why was disregarded since the last review. I'd suggest having a thorough look and reply to every unresolved conversations to make sure we all align on the directions.

I think we are pretty close to completion, WDYT @isuckatcs?

fangyi-zhou commented:
Thanks for the review. I might have missed some comments since I was away from this pull request for quite a while and I probably forgot. I'll have another revision.

fangyi-zhou commented on clang/include/clang/StaticAnalyzer/Checkers/SValExplainer.h:
```diff
@@ -114,7 +128,8 @@ class SValExplainer : public FullSValVisitor<SValExplainer, std::string> {
   std::string VisitSymbolConjured(const SymbolConjured *S) {
     return "symbol of type '" + S->getType().getAsString() +
-           "' conjured at statement '" + printStmt(S->getStmt()) + "'";
+           "' conjured at statement '" +
```
My intention was to keep this is as for now to avoid making visible changes (i.e. the emitted diagnostics). If it's desirable to change this to a CFG Element, I can make the required change (although that would include changing some test cases)

steakhal replied:
Please do. These are only used in tests, and not in production.

fangyi-zhou committed:
Address most of the reviewer comments

Test are expected to fail, to be fixed in a later commit.

fangyi-zhou committed:
Fix conjured symbol visiter printout

fangyi-zhou committed:
Fix loop widening element ref

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2556,10 +2556,19 @@ void ExprEngine::processCFGBlockEntrance(const BlockEdge &L,
     if (!isa_and_nonnull<ForStmt, WhileStmt, DoStmt, CXXForRangeStmt>(Term))
       return;
+
+    // FIXME:
```
Please have a look at this workaround. I've noticed that, after adding back the assertion, the CFG element from the `ExprEngine` was incorrect. Therefore this workaround is put in place as a replacement for the terminator.

steakhal replied:
Now thinking about this, and looking at the assertion within `getWidenedLoopState` we don't actually need to assert that the wrapped Stmt is a loop statement. As for correctness, it might be surprising to hear but it also doesn't really matter what CFGElement we would pass, the important part to pass something, and do so deterministically.

So if it makes the code simpler, we can drop the assertion inside, and even pass the wrong CFGElement here.
Would it make the patch simpler?

fangyi-zhou replied:
Fair enough, I will remove the assertion but leave this FIXME here for future reference.

fangyi-zhou replied:
I suspect we need to actually provide a valid CFG element here, so maybe we need to pass the first element in the new CFG Block?

fangyi-zhou committed:
Remove previous workaround

fangyi-zhou commented:
Looks like there's an Asan issue at the stale CFG element for loop widening https://lab.llvm.org/buildbot/#/builders/55/builds/10398. Do we need to revert?

steakhal commented:
@fangyi-zhou Could you please check what's wrong?
PS: I reverted this patch for now.
